### ISSUE TITLE
Patch literal JSON in ConfigMap item

### ISSUE THREAD

**Issue by narg95** [kubernetes-sigs/kustomize#680]
Hi all,

I want to patch a literal json from a configMap item, but I have not found an example in the docu,  do you know if that is possible? For example, assuming the following configmap:

Base configMap
```sh
apiVersion: v1
kind: ConfigMap
metadata:
  name: demo-configmap
data:   
   settings:  |-
      {
         foo: "foo_1",
         bar: "bar_1"
         ...  // assume more attributes
      }
   other-settings: |-
      { 
          ...
      }
```
I want to apply this patch:

```sh
apiVersion: v1
kind: ConfigMap
metadata:
  name: demo-configmap
data:   
settings:  |-
   {
         bar: "bar_2"
   }
```

So the resulting configMap looks like:

```sh
apiVersion: v1
kind: ConfigMap
metadata:
  name: demo-configmap
data:   
   settings:  |-
      {
         foo: "foo_1",
         bar: "bar_2" // Here got patched !
         ...  // all the other attributes remain
      }
   other-settings: |-
      { 
          ... // same here
      }
```

Thanks !

P.S. Great job with kustomize!

**Comment by Liujingfang1** [comment]
@narg95 Patching doesn't support that. It will replace the value of `settings` rather than merging the values. 
I'd suggest use [ConfigMapGenerator](https://github.com/kubernetes-sigs/kustomize/blob/master/docs/kustomization.yaml#L79) from files. For different base or overlays, you can adjust different the file content.

**Comment by ajbouh** [comment]
I need this as well.

Another option would be to pass along var references to the external program as environment variables and use the output of that program as the configmap. @Liujingfang1, would that be better aligned with the philosophy of kustomize?

**Comment by narg95** [comment]
Hi @Liujingfang1,

> I'd suggest use [ConfigMapGenerator](https://github.com/kubernetes-sigs/kustomize/blob/master/docs/kustomization.yaml#L79) from files. For different base or overlays, you can adjust different the file content.

In a greenfield project I will do as suggested, but I am porting an application, that uses a rather complex json config file, to kubernetes, so using a ConfigMapGenerator will not be ideal because I would have to repeate the whole JSON file for a trivial json property change. I currently solve it post-processing kustomize's output with `envsubst` but this is something I want to avoid therefore it would be great that kustomize would help me to do so. 

If this is a wanted feature I am willing to invest some time in a PR, but if this is something very specific to my use-case I rather solve it by using configMap literals, which can be re-used and overrided, and add a sidecar that converts it to the expected application config json format.

@Liujingfang1  let me know your thoughts!

**Comment by Liujingfang1** [comment]
@narg95 According to the comment history here, it could be a common use case. If would be nice if Kustomize can provide some convenience here.

The example listed is a JSON file as the value. There are other cases as well. For example, I may have the configmap generated from following files
```
key1=value1
key2=value2
```
There is no limitation on what type of files can be used to create a ConfigMap. How to detect the format and apply proper merging is tricky.

**Comment by narg95** [comment]
@Liujingfang1 thanks for your feedback!

> I may have the configmap generated from following files

did you mean :*generated from a file with the following values*, right?

> There is no limitation on what type of files can be used to create a ConfigMap.

right, I think the file type support should extensible and we can start with JSON, then props (key=value) and then perhaps yaml.

I was thinking in how it could look like, here are my two cetns on the syntax:

base config map 
```
configMapGenerator:
name: demo-settings
- file: my-app-settings.json
- file: foo.json
- file: bar.json
```

my-app-settings.json
```json
{
    "name": "my app",
    "environment": "demo",
    "version": "1.0",
    "args:" ["--debug-mode"]
}
```

---
Overlay

kustomization.yaml
```
-base: ../base/my-app/

configMapGenerator:
name: demo-settings
- patch: my-app-settings.json # file name gives us the item name and file extension give us the file type and therefore the default json merge strategy. 
```

```
{
    "version": "2.0-rc1",
    "args:" ["--debug-mode", "off"]
}
```
`kustomize build` output for configMap

```
apiVersion: v1
kind: ConfigMap
metadata:
  name: demo-settings
data:   
   my-app-settings.json:  |-
      {
        "name": "my app",
        "environment": "demo",
        "version": "2.0-rc1",  // PATCH
        "args:" ["--debug-mode", "off"] // PATCH
      }
```

---

The `- patch` attribute supports a short and long version in order to make it extensible, 

short version
```
configMapGenerator:
name: demo-settings
- patch: my-app-settings.json 
```
The file name give us the configmap item key to patch, the file extension gives us the patch type  `*.json: strategic-merge`, `*.json6902: json patch 6902`, etc

long version
```
configMapGenerator:
name: demo-settings
- patch:
  item: my-app-settings.json. # matches config map item key
  path: patches/patch.json 
  type: strategic-merge # it can be any of [json-6902,yaml-6902, props, etc..]  
```

opinions :) ?

**Comment by ikatson** [comment]
One more option to discuss here: make it possible to render configmap/secret content as templates, e.g. golang templates.

I understand that kustomize's awesomeness stems from the fact that it does not use templating to build kubernetes objects which are structured data by nature. But configmap/secret content is opaque data, so making it possible to render with templates sounds cool.

For example make it work like helm template processing, but only for configmap/secret content. Maybe other literal field content, e.g. JSON stored in annotations.

Here's a **very rough** sketch of how this might look like:

```
#### base/app.properties
foo={{ .Values.foo }}
bar={{ .Values.bar }}

#### base/values-defaults.yaml
foo: 'foo-value'
bar: 'bar-value'

#### base/kustomization.yaml
valueSources:
- values-defaults.yaml
  
configMapGenerator
- name: myJavaServerProps
  render:
  # the values for rendering app.properties will come from "valueSources", which might get enriched in overlays
  - app.properties

#### overlays/production/production-values.yaml
bar: 'bar-value-production'

#### overlays/production/kustomization.yaml
valueSources: 
- production-values.yaml

bases:
# the base gets rendered using combined value sources: "production-values.yaml" overlaid on top of "base/values-defaults.yaml"
- ../base/
```

I'm not presenting an API for review here as it definitely needs more thought, but more an idea of what you could accomplish with it.

**Comment by monopole** [comment]
Attemping to summarize:

Say you want to use kustomize to make the configmaps used in production differ in some way from that used in development and staging.

Mixins might work.  Suppose the configmap held a java properties file (k=v pairs, one per line), and one wanted to use kustomize to “edit” the value for one particular key.  The kustomize way to do this would be to split the file - put the bulk of the k=v’s into a configmap in the base, then put the particular k=v in an overlay and get the final configmap as a merger ([example](https://github.com/kubernetes-sigs/kustomize/blob/master/examples/combineConfigs.md)) which then goes to production or development, etc.  This works because kustomize understands that particular file format.  It knows the values are whatever comes between the `=` and EOL characters.

More generally, configmap values are opaque to k8s - they could be a simple port number, or some snippet of json/javascript/lisp/whatever.

A scheme to allow kustomize to edit _any_ configmap value would have to be in the form of an _unstructured edit_, e.g. 

 * for configmap _foo_, look up the value for key _bar_, and run this regexp replacement pattern over it.
 * Or - as @ikatson suggests - the final value for _bar_ is the expansion of some Go template, and we get different variants by specifying different template mappings in different overlays.  This is the helm approach, but just for map values instead of everything in sight.

@narg95 focusses on a structured JSON edit instead of an unstructure edit, to address the particular case of an app being ported to k8s that consumes some complex JSON config file, and @narg95 wants to apply structured edits (e.g. to get loglevel:info in production instead of loglevel:debug).

@narg95 SGTM - can you write a short KEP (e.g [892](https://github.com/kubernetes/enhancements/pull/892),  [886](https://github.com/kubernetes/enhancements/pull/886)) - you pretty much have the meat of it above -   and then provide the impl?  Do you have any notion of how common this might be?  It’s a structured edit, so it’s consistent with kustomize in that sense.

As an aside, Secrets are getting a big upgrade, where we allow a plugin to generate KV pairs (since the V’s are not just opaque - they are _secret_).  This isn’t an patch mechanism (create KV’s only), and it’s not going to be active in configmaps - but folks on this thread might be interested.

**Comment by yellowmegaman** [comment]
@monopole Maybe I'm missing something, but adding another file to configMap is not actually a merge.
What  we do want (speaking for my company) is to have some basic common application.conf, and some custom additions to it, like dev.conf/stage.conf which are ok to just to append to common config.
With your example https://github.com/kubernetes-sigs/kustomize/blob/master/examples/combineConfigs.md we'll achieve two separate config files, which are not supported at the moment.

**Comment by matti** [comment]
this is another case where variables would help like https://github.com/kontena/mortar has

**Comment by kkasravi** [comment]
@narg95 this can be done in the following way:

Below is in a file `resources.yaml`

```
apiVersion: v1
kind: ConfigMap
metadata:
  name: demo-configmap
data:   
   settings:  |-
      {
         foo: $(foo),
         bar: $(bar)
      }
```

The kustomization file looks like 

```
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization
resources:
- resources.yaml
configMapGenerator:
- name: demo-configmap-parameters
  env: params.env
generatorOptions:
  disableNameSuffixHash: true
vars:
- name: foo
  objref:
    kind: ConfigMap
    name: demo-configmap-parameters
    apiVersion: v1
  fieldref:
    fieldpath: data.foo
- name: bar
  objref:
    kind: ConfigMap
    name: demo-configmap-parameters
    apiVersion: v1
  fieldref:
    fieldpath: data.bar
configurations:
- params.yaml
```

params.env contains

```
foo=foo_1
bar=bar_2
```

params.yaml contains

```
varReference:
- path: data/settings
  kind: ConfigMap
```

Note that this inserts an extra ConfigMap which you would probably delete afterwards.
eg

```
$ ☞  kubectl create ns demo 1>/dev/null && kustomize build | kubectl apply -f - -n demo 1>/dev/null && kubectl delete cm demo-configmap-parameters -n demo 1>/dev/null && kubectl get cm demo-configmap -oyaml -n demo && kubectl delete ns demo 1>/dev/null
```

outputs
```
apiVersion: v1
data:
  settings: |-
    {
       foo: "foo_1",
       bar: "bar_2"
    }
kind: ConfigMap
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"v1","data":{"settings":"{\n   foo: \"foo_1\",\n   bar: \"bar_2\"\n}"},"kind":"ConfigMap","metadata":{"annotations":{},"name":"demo-configmap","namespace":"demo"}}
  creationTimestamp: "2019-04-07T14:47:54Z"
  name: demo-configmap
  namespace: demo
  resourceVersion: "40657790"
  selfLink: /api/v1/namespaces/demo/configmaps/demo-configmap
  uid: 203db6aa-5944-11e9-9a72-42010a8a016c
```

**Comment by houz42** [comment]
@kkasravi thanks for your examples, but do we have a more detailed documentation for this workflow? I want to generate a configMap from a nginx config file, with some template values using the 'vars'. But theres is too much '$' signs in it, which should be treated as literals. Is there any way to tell kustomize not to render them.

**Comment by houz42** [comment]
by the way, kustomize will output as follows if there is any unresolable '$' signs
```
Error: couldn't make target for ../../base: json: cannot unmarshal object into Go struct field Kustomization.configMapGenerator of type []types.ConfigMapArgs
```

**Comment by narg95** [comment]
Hi @kkasravi, thanks for the explanation and the code!

I ended up with a similar solution but in my case at that time I generated all the configs with `kustomize config save -d ./configs` and I appended my custom configuration including the path `data/setings` from  configmaps. I did it so because from my understanding the custom configurations did not get merged with defaults so I did not wanted to lose kustomize's default config. But in your example you just added a new configuration so I am wondering if you get the defaults plus the ones defined in your `params.yml` file.

**Comment by benjamin-bergia** [comment]
Hi, just met the same issue here.
Without going into templating or "smart" merging, being able to concatenate multiple file under the same key in the config map would make a huge difference. I think the current syntax is neat and doesn't have to change but having (as in the example):
`base/kustomization.yml`
```
configMapGenerator:
- name: my-configmap
  files:
  - settings.cfg
```
`overlays/prod/kustomization.yml`
```
configMapGenerator:
- name: my-configmap
  behavior: merge
  files:
  - settings.cfg
```
Should result in a configmap with a single key `settings.cfg` and the content of both files concatenated.
It is not optimal since it requires using the same file names but would still be a big improvement in maintainability.

**Comment by SilverXXX** [comment]
Hi, same problem here.
@benjamin-bergia solution would be enough even for us, solving problems when multiple files cannot be loaded.

**Comment by jbrette** [comment]
@SilverXXX @benjamin-bergia The configMapGenerator overlay does not seems to work write. There is bug in the Resource.replace/merge implementation and we end up with invalid labels: 

```yaml
apiVersion: v1
data:
  settings.cfg: |
    Content of the overlay settings.cfg
kind: ConfigMap
metadata:
  annotations: {}
  labels: {}
  name: my-configmap-8gkkcddg96
```

**Comment by jbrette** [comment]
@ikatson @monopole @Liujingfang1 I think this another use case where the [autovar PR](https://github.com/kubernetes-sigs/kustomize/pull/1217)

We are basically able to achieve exactly what ikatson was asking without templating and by levering multiple key features of kustomize.  Have a look at the [third solution](https://github.com/keleustes/kustomize/tree/allinone/examples/issues/issue_0680#verification-step-expected2)

The output is basically the following:
```yaml
apiVersion: v1
data:
  app.properties: |
    foo=production-foo-value
    bar=production-foo-value
kind: ConfigMap
metadata:
  name: myJavaServerProps-7m6dhd66bg
---
apiVersion: v1
kind: Values
metadata:
  name: file1
spec:
  bar: production-foo-value
  foo: production-foo-value
```

**Comment by nthypes** [comment]
Nothing? 

Sometimes the unresponsive attitude of kustomize team about open issues/prs it's really sad.

**Comment by davisford** [comment]
Will there ever be a valid solution for patching a JSON `ConfigMap`?  The lame workaround right now is to copy the full JSON file into every overlay and manually override it like:

base/thing/kustomization.yaml
```
configMapGenerator:
 - name: thing-config
    files: 
        - config/thing-config.json
```

overlay/foo/kustomization.yaml
```
configMapGenerator:
  - name: thing-config
    behavior: merge
    files:
      - thing-config.json
```

This is really error prone and not ideal.  Is there not _some_ way we can use `patchesJson6902` here to patch a JSON file?  That's what it was intended for, but so far I've yet to work out a config that works with this kind of setup.

**Comment by liger1978** [comment]
Just a note to say this is still needed...

**Comment by afirth** [comment]
not sure if i can/should but
/reopen
the ability to patch structured configmaps is such a pain point in older apps, particularly in javaland.

**Comment by matti** [comment]
Open a new issue

**Comment by goto100** [comment]
Can't imaging there is still no resolution for this in 2021……

**Comment by m62534** [comment]
How do we do this now that var is deprecated?
I don't think varReference works anymore either when we replace with `replacements`
